# ECT_HW5
### 1

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

df = pd.read_csv('customer_churn.csv')
df.head(10)

CustomerID  Churn  Tenure PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001      1     4.0         Mobile Phone         3              6.0   
1       50002      1     NaN                Phone         1              8.0   
2       50003      1     NaN                Phone         1             30.0   
3       50004      1     0.0                Phone         3             15.0   
4       50005      1     0.0                Phone         1             12.0   
5       50006      1     0.0             Computer         1             22.0   
6       50007      1     NaN                Phone         3             11.0   
7       50008      1     NaN                Phone         1              6.0   
8       50009      1    13.0                Phone         3              9.0   
9       50010      1     NaN                Phone         1             31.0   

  PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0           Debit Card  Female             3.0                         3   
1                  UPI    Male             3.0                         4   
2           Debit Card    Male             2.0                         4   
3           Debit Card    Male             2.0                         4   
4                   CC    Male             NaN                         3   
5           Debit Card  Female             3.0                         5   
6     Cash on Delivery    Male             2.0                         3   
7                   CC    Male             3.0                         3   
8             E wallet    Male             NaN                         4   
9           Debit Card    Male             2.0                         5   

     PreferedOrderCat  SatisfactionScore MaritalStatus  NumberOfAddress  \
0  Laptop & Accessory                  2        Single                9   
1              Mobile                  3        Single                7   
2              Mobile                  3        Single                6   
3  Laptop & Accessory                  5        Single                8   
4              Mobile                  5        Single                3   
5        Mobile Phone                  5        Single                2   
6  Laptop & Accessory                  2      Divorced                4   
7              Mobile                  2      Divorced                3   
8              Mobile                  3      Divorced                2   
9              Mobile                  3        Single                2   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   
5         1                         22.0         4.0         6.0   
6         0                         14.0         0.0         1.0   
7         1                         16.0         2.0         2.0   
8         1                         14.0         0.0         1.0   
9         0                         12.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  
0                5.0             160  
1                0.0             121  
2                3.0             120  
3                3.0             134  
4                3.0             130  
5                7.0             139  
6                0.0             121  
7                0.0             123  
8                2.0             127  
9                1.0             123

In [2]:
#將字串轉換成數字分類

#PreferredLoginDevice部分
df.loc[df.PreferredLoginDevice == 'Computer', 'PreferredLoginDevice'] = 0
df.loc[df.PreferredLoginDevice == 'Mobile Phone', 'PreferredLoginDevice'] = 1
df.loc[df.PreferredLoginDevice == 'Phone', 'PreferredLoginDevice'] = 2

#PreferredPaymentMode部分
df.loc[df.PreferredPaymentMode == 'CC', 'PreferredPaymentMode'] = 0
df.loc[df.PreferredPaymentMode == 'COD', 'PreferredPaymentMode'] = 1
df.loc[df.PreferredPaymentMode == 'Cash on Delivery', 'PreferredPaymentMode'] = 2
df.loc[df.PreferredPaymentMode == 'Credit Card', 'PreferredPaymentMode'] = 3
df.loc[df.PreferredPaymentMode == 'Debit Card', 'PreferredPaymentMode'] = 4
df.loc[df.PreferredPaymentMode == 'E wallet', 'PreferredPaymentMode'] = 5
df.loc[df.PreferredPaymentMode == 'UPI', 'PreferredPaymentMode'] = 6

#Gender部分
df.loc[df.Gender == 'Male', 'Gender'] = 0
df.loc[df.Gender == 'Female', 'Gender'] = 1

#PreferedOrderCat部分
df.loc[df.PreferedOrderCat == 'Fashion', 'PreferedOrderCat'] = 0
df.loc[df.PreferedOrderCat == 'Grocery', 'PreferedOrderCat'] = 1
df.loc[df.PreferedOrderCat == 'Laptop & Accessory', 'PreferedOrderCat'] = 2
df.loc[df.PreferedOrderCat == 'Mobile', 'PreferedOrderCat'] = 3
df.loc[df.PreferedOrderCat == 'Mobile Phone', 'PreferedOrderCat'] = 4
df.loc[df.PreferedOrderCat == 'Others', 'PreferedOrderCat'] = 5

#MaritalStatus部分
df.loc[df.MaritalStatus == 'Divorced', 'MaritalStatus'] = 0
df.loc[df.MaritalStatus == 'Married', 'MaritalStatus'] = 1
df.loc[df.MaritalStatus == 'Single', 'MaritalStatus'] = 2

df.head(10)

CustomerID  Churn  Tenure  PreferredLoginDevice  CityTier  WarehouseToHome  \
0       50001      1     4.0                     1         3              6.0   
1       50002      1     NaN                     2         1              8.0   
2       50003      1     NaN                     2         1             30.0   
3       50004      1     0.0                     2         3             15.0   
4       50005      1     0.0                     2         1             12.0   
5       50006      1     0.0                     0         1             22.0   
6       50007      1     NaN                     2         3             11.0   
7       50008      1     NaN                     2         1              6.0   
8       50009      1    13.0                     2         3              9.0   
9       50010      1     NaN                     2         1             31.0   

   PreferredPaymentMode  Gender  HourSpendOnApp  NumberOfDeviceRegistered  \
0                     4       1             3.0                         3   
1                     6       0             3.0                         4   
2                     4       0             2.0                         4   
3                     4       0             2.0                         4   
4                     0       0             NaN                         3   
5                     4       1             3.0                         5   
6                     2       0             2.0                         3   
7                     0       0             3.0                         3   
8                     5       0             NaN                         4   
9                     4       0             2.0                         5   

   PreferedOrderCat  SatisfactionScore  MaritalStatus  NumberOfAddress  \
0                 2                  2              2                9   
1                 3                  3              2                7   
2                 3                  3              2                6   
3                 2                  5              2                8   
4                 3                  5              2                3   
5                 4                  5              2                2   
6                 2                  2              0                4   
7                 3                  2              0                3   
8                 3                  3              0                2   
9                 3                  3              2                2   

   Complain  OrderAmountHikeFromlastYear  CouponUsed  OrderCount  \
0         1                         11.0         1.0         1.0   
1         1                         15.0         0.0         1.0   
2         1                         14.0         0.0         1.0   
3         0                         23.0         0.0         1.0   
4         0                         11.0         1.0         1.0   
5         1                         22.0         4.0         6.0   
6         0                         14.0         0.0         1.0   
7         1                         16.0         2.0         2.0   
8         1                         14.0         0.0         1.0   
9         0                         12.0         1.0         1.0   

   DaySinceLastOrder  CashbackAmount  
0                5.0             160  
1                0.0             121  
2                3.0             120  
3                3.0             134  
4                3.0             130  
5                7.0             139  
6                0.0             121  
7                0.0             123  
8                2.0             127  
9                1.0             123

In [3]:
#空值部分直接以平均數帶入
df.fillna(df.mean(),inplace=True)
df.head(10)

CustomerID  Churn    Tenure  PreferredLoginDevice  CityTier  \
0       50001      1   4.00000                     1         3   
1       50002      1  10.18657                     2         1   
2       50003      1  10.18657                     2         1   
3       50004      1   0.00000                     2         3   
4       50005      1   0.00000                     2         1   
5       50006      1   0.00000                     0         1   
6       50007      1  10.18657                     2         3   
7       50008      1  10.18657                     2         1   
8       50009      1  13.00000                     2         3   
9       50010      1  10.18657                     2         1   

   WarehouseToHome  PreferredPaymentMode  Gender  HourSpendOnApp  \
0              6.0                     4       1        3.000000   
1              8.0                     6       0        3.000000   
2             30.0                     4       0        2.000000   
3             15.0                     4       0        2.000000   
4             12.0                     0       0        2.930734   
5             22.0                     4       1        3.000000   
6             11.0                     2       0        2.000000   
7              6.0                     0       0        3.000000   
8              9.0                     5       0        2.930734   
9             31.0                     4       0        2.000000   

   NumberOfDeviceRegistered  PreferedOrderCat  SatisfactionScore  \
0                         3                 2                  2   
1                         4                 3                  3   
2                         4                 3                  3   
3                         4                 2                  5   
4                         3                 3                  5   
5                         5                 4                  5   
6                         3                 2                  2   
7                         3                 3                  2   
8                         4                 3                  3   
9                         5                 3                  3   

   MaritalStatus  NumberOfAddress  Complain  OrderAmountHikeFromlastYear  \
0              2                9         1                         11.0   
1              2                7         1                         15.0   
2              2                6         1                         14.0   
3              2                8         0                         23.0   
4              2                3         0                         11.0   
5              2                2         1                         22.0   
6              0                4         0                         14.0   
7              0                3         1                         16.0   
8              0                2         1                         14.0   
9              2                2         0                         12.0   

   CouponUsed  OrderCount  DaySinceLastOrder  CashbackAmount  
0         1.0         1.0                5.0             160  
1         0.0         1.0                0.0             121  
2         0.0         1.0                3.0             120  
3         0.0         1.0                3.0             134  
4         1.0         1.0                3.0             130  
5         4.0         6.0                7.0             139  
6         0.0         1.0                0.0             121  
7         2.0         2.0                0.0             123  
8         0.0         1.0                2.0             127  
9         1.0         1.0                1.0             123

In [4]:
sample = df.groupby('Churn', group_keys=False).apply(lambda x: x.sample(frac=0.6))
sample

CustomerID  Churn    Tenure  PreferredLoginDevice  CityTier  \
5271       55262      0   6.00000                     1         1   
3602       53595      0  22.00000                     1         1   
5061       55052      0  18.00000                     2         1   
399        50392      0   0.00000                     2         1   
1039       51032      0  13.00000                     1         1   
1362       51355      0   9.00000                     1         1   
977        50970      0  13.00000                     1         3   
2088       52081      0  21.00000                     2         1   
4909       54900      0   1.00000                     1         1   
4960       54951      0   7.00000                     1         1   
443        50436      0   7.00000                     1         1   
2714       52707      0   0.00000                     0         1   
3949       53942      0   4.00000                     1         3   
2148       52141      0  10.00000                     1         3   
483        50476      0   5.00000                     1         1   
3817       53810      0  12.00000                     1         3   
3282       53275      0  16.00000                     0         1   
4311       54304      0   4.00000                     1         1   
1827       51820      0   0.00000                     2         1   
215        50209      0  16.00000                     1         1   
5321       55312      0   2.00000                     1         1   
3147       53140      0  13.00000                     0         1   
2560       52553      0  28.00000                     2         1   
2719       52712      0  12.00000                     0         3   
4452       54445      0   1.00000                     0         1   
3399       53392      0  27.00000                     1         3   
4512       54505      0   9.00000                     1         1   
5190       55181      0  21.00000                     1         1   
3878       53871      0  10.00000                     2         1   
3633       53626      0   6.00000                     1         3   
...          ...    ...       ...                   ...       ...   
4495       54488      1   3.00000                     1         3   
3808       53801      1   1.00000                     2         1   
2590       52583      1   0.00000                     1         3   
18         50019      1   0.00000                     0         1   
4648       54641      1   1.00000                     0         3   
3283       53276      1   1.00000                     2         1   
889        50882      1   0.00000                     0         1   
5182       55173      1   1.00000                     0         3   
4461       54454      1   1.00000                     1         3   
2410       52403      1   0.00000                     0         3   
1173       51166      1   0.00000                     0         1   
4759       54752      1   1.00000                     1         3   
3756       53749      1   1.00000                     1         3   
2880       52873      1   5.00000                     1         3   
2903       52896      1   1.00000                     1         3   
1050       51043      1   0.00000                     1         3   
1353       51346      1   4.00000                     1         3   
1460       51453      1   0.00000                     1         3   
92         50092      1   0.00000                     0         3   
5383       55374      1   1.00000                     2         3   
52         50053      1   0.00000                     1         1   
2515       52508      1   0.00000                     1         3   
1129       51122      1  10.18657                     0         2   
794        50787      1   0.00000                     0         3   
3664       53657      1  21.00000                     1         1   
1252       51245      1   0.00000                     1         3   
939 

### 2

In [5]:
sample['Churn'].value_counts()

0    2813
1     571
Name: Churn, dtype: int64

In [6]:
sample.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3384 entries, 5271 to 1646
Data columns (total 20 columns):
CustomerID                     3384 non-null int64
Churn                          3384 non-null int64
Tenure                         3384 non-null float64
PreferredLoginDevice           3384 non-null int64
CityTier                       3384 non-null int64
WarehouseToHome                3384 non-null float64
PreferredPaymentMode           3384 non-null int64
Gender                         3384 non-null int64
HourSpendOnApp                 3384 non-null float64
NumberOfDeviceRegistered       3384 non-null int64
PreferedOrderCat               3384 non-null int64
SatisfactionScore              3384 non-null int64
MaritalStatus                  3384 non-null int64
NumberOfAddress                3384 non-null int64
Complain                       3384 non-null int64
OrderAmountHikeFromlastYear    3384 non-null float64
CouponUsed                     3384 non-null float64
OrderCount   

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5640 entries, 0 to 5639
Data columns (total 20 columns):
CustomerID                     5640 non-null int64
Churn                          5640 non-null int64
Tenure                         5640 non-null float64
PreferredLoginDevice           5640 non-null int64
CityTier                       5640 non-null int64
WarehouseToHome                5640 non-null float64
PreferredPaymentMode           5640 non-null int64
Gender                         5640 non-null int64
HourSpendOnApp                 5640 non-null float64
NumberOfDeviceRegistered       5640 non-null int64
PreferedOrderCat               5640 non-null int64
SatisfactionScore              5640 non-null int64
MaritalStatus                  5640 non-null int64
NumberOfAddress                5640 non-null int64
Complain                       5640 non-null int64
OrderAmountHikeFromlastYear    5640 non-null float64
CouponUsed                     5640 non-null float64
OrderCount      

### 3, 4
前處理已在前面處理完

In [8]:
import copy
from sklearn import svm
from sklearn.svm import SVR
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

#from sklearn.cross_validation import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score


X = df
y = df['Churn']

X_train = sample.drop(['Churn'], axis = 1)
y_train = sample['Churn']#要預測的欄位

X_test = copy.deepcopy(df[~df.index.isin(sample.index.values)])
y_test = df['Churn'].drop(sample['Churn'].index)

In [9]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2256 entries, 0 to 5637
Data columns (total 20 columns):
CustomerID                     2256 non-null int64
Churn                          2256 non-null int64
Tenure                         2256 non-null float64
PreferredLoginDevice           2256 non-null int64
CityTier                       2256 non-null int64
WarehouseToHome                2256 non-null float64
PreferredPaymentMode           2256 non-null int64
Gender                         2256 non-null int64
HourSpendOnApp                 2256 non-null float64
NumberOfDeviceRegistered       2256 non-null int64
PreferedOrderCat               2256 non-null int64
SatisfactionScore              2256 non-null int64
MaritalStatus                  2256 non-null int64
NumberOfAddress                2256 non-null int64
Complain                       2256 non-null int64
OrderAmountHikeFromlastYear    2256 non-null float64
CouponUsed                     2256 non-null float64
OrderCount      

In [10]:
y_test

0       1
1       1
2       1
4       1
6       1
7       1
10      1
14      1
22      1
23      1
25      1
27      0
29      0
30      0
33      0
35      0
38      0
39      0
40      0
42      0
43      0
45      0
46      1
47      0
49      0
50      0
51      0
54      0
57      1
59      0
       ..
5563    0
5566    1
5570    0
5573    0
5574    0
5576    0
5578    0
5579    0
5585    0
5590    0
5594    0
5596    0
5599    0
5601    0
5603    0
5607    0
5616    0
5618    0
5619    0
5622    1
5623    0
5624    0
5627    0
5628    0
5631    1
5632    0
5633    0
5635    0
5636    0
5637    0
Name: Churn, Length: 2256, dtype: int64

logistic regression 部分

In [11]:
lmodel = LogisticRegression()
lmodel.fit(X_train, y_train)

lscores = cross_val_score(lmodel,X_test,y_test,cv=10,scoring='accuracy')
print(lscores)
print(lscores.mean())

[0.8539823  0.84513274 0.8539823  0.87168142 0.8539823  0.87610619
 0.84888889 0.87555556 0.88444444 0.85777778]
0.8621533923303832


SVM部分

In [ ]:
# #kernel = linear
# svmModel = svm.SVR(C=1,kernel='linear') 
# svmModel.fit(X_train, y_train)

# svmscores = cross_val_score(svmModel,X_test,y_test,cv=10,scoring='accuracy')
# print(svmscores)
# print(svmscores.mean())

In [ ]:
# Model=svm.SVC(kernel='linear',C=1)
# Model.fit(X_train, y_train)

# Cscores = cross_val_score(Model,X_test,y_test,cv=10,scoring='accuracy')
# print(Cscores)
# print(Cscores.mean())

In [12]:
svcModel = svm.LinearSVC(C=1) 
svcModel.fit(X_train, y_train)

cscores = cross_val_score(svcModel,X_test,y_test,cv=10,scoring='accuracy')
print(cscores)
print(cscores.mean())

[0.83185841 0.83185841 0.83185841 0.16814159 0.83185841 0.83185841
 0.83111111 0.83111111 0.83111111 0.83111111]
0.765187807276303


### 5

In [13]:
lscrsum = 0.0
sscrsum = 0.0
lscrArr = []
sscrArr = []

for i in range(1,30):
    sample = df.groupby('Churn', group_keys=False).apply(lambda x: x.sample(frac=0.6))
    X = df
    y = df['Churn']

    X_train = sample.drop(['Churn'], axis = 1)
    y_train = sample['Churn']#要預測的欄位

    X_test = copy.deepcopy(df[~df.index.isin(sample.index.values)])
    y_test = df['Churn'].drop(sample['Churn'].index)
    
    print("第",i,"次：")
    
    
    #logistic regression
    print("logistic regression:")
    lmodel = LogisticRegression()
    lmodel.fit(X_train, y_train)

    lscores = cross_val_score(lmodel,X_test,y_test,cv=10,scoring='accuracy')
    print(lscores.mean())
    lscrsum+=lscores.mean()
    lscrArr.append(lscores.mean())
    
    
    #SVM
    print("SVM:")
    svcModel = svm.LinearSVC(C=1) 
    svcModel.fit(X_train, y_train)

    cscores = cross_val_score(svcModel,X_test,y_test,cv=10,scoring='accuracy')
    print(cscores.mean())
    sscrsum+=cscores.mean()
    sscrArr.append(cscores.mean())
    
    #Model=svm.SVC(kernel='linear',C=1)
    #Model.fit(X_train, y_train)

    #Cscores = cross_val_score(Model,X_test,y_test,cv=10,scoring='accuracy')
    #print(Cscores.mean())
    #sscrsum+=Cscores.mean()
    
print("logistic regression最終的平均accuracy:",lscrsum/30)
print("SVM最終的平均accuracy:",sscrsum/30)

第 1 次：
logistic regression:
0.861683382497542
SVM:
0.8315594886922322
第 2 次：
logistic regression:
0.857732546705998
SVM:
0.765187807276303
第 3 次：
logistic regression:
0.8754670599803344
SVM:
0.6988161258603738
第 4 次：
logistic regression:
0.8581671583087512
SVM:
0.7653372664700099
第 5 次：
logistic regression:
0.8674690265486726
SVM:
0.6988161258603738
第 6 次：
logistic regression:
0.8674886922320549
SVM:
0.8315594886922322
第 7 次：
logistic regression:
0.8670462143559489
SVM:
0.6991150442477876
第 8 次：
logistic regression:
0.8692527040314652
SVM:
0.7653372664700099
第 9 次：
logistic regression:
0.8630344149459195
SVM:
0.8315594886922322
第 10 次：
logistic regression:
0.8692487708947887
SVM:
0.7653372664700099
第 11 次：
logistic regression:
0.8643618485742379
SVM:
0.8315594886922322
第 12 次：
logistic regression:
0.8798721730580137
SVM:
0.6989655850540807
第 13 次：
logistic regression:
0.8501848574237956
SVM:
0.8315594886922322
第 14 次：
logistic regression:
0.8572566371681415
SVM:
0.7653372664700099
第 15

### 6

In [14]:
import scipy.stats as stats

stats.ttest_rel(lscrArr, sscrArr)

Ttest_relResult(statistic=9.975213492478945, pvalue=1.0166554502873033e-10)

The test statistic comes out to be equal to 9.975 and the corresponding two-sided p-value is 1.017e-10.


The paired samples t-test follows the null and alternative hypotheses:

H0: 表示兩個model的Accuracy相等<br>
HA: 表示兩個model的Accuracy不相等

As the p-value comes out to be equal to 1.017e-10 which is less than 0.05 hence we reject the null hypothesis. So, we have enough proof to claim that the accuracy score of two models is different. 